pkg/api/util/clusteragent.go

```diff
@@ -0,0 +1,185 @@
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+	log "github.com/cihub/seelog"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
+)
+
+const (
+	authorizationHeaderKey      = "Authorization"
+	clusterAgentAuthTokenMinLen = 32
+)
+
+var globalClusterAgentUtil *ClusterAgentUtil
+
+type serviceNames []string
+
+// ClusterAgentUtil is required to query the API of Datadog cluster agent
+type ClusterAgentUtil struct {
+	// used to setup the ClusterAgentUtil
+	initRetry retry.Retrier
+
+	clusterAgentAPIEndpoint       string // ${SCHEME}://${clusterAgentHost}:${PORT}
+	clusterAgentAPIClient         *http.Client
+	clusterAgentAPIRequestHeaders *http.Header
+}
+
+// GetClusterAgentUtil returns or init the ClusterAgentUtil
+func GetClusterAgentUtil() (*ClusterAgentUtil, error) {
+	if globalClusterAgentUtil == nil {
+		globalClusterAgentUtil = &ClusterAgentUtil{}
+		globalClusterAgentUtil.initRetry.SetupRetrier(&retry.Config{
+			Name:          "clusterAgentClient",
+			AttemptMethod: globalClusterAgentUtil.init,
+			Strategy:      retry.RetryCount,
+			RetryCount:    10,
+			RetryDelay:    30 * time.Second,
+		})
+	}
+	err := globalClusterAgentUtil.initRetry.TriggerRetry()
+	if err != nil {
+		log.Debugf("Init error: %s", err)
+		return nil, err
+	}
+	return globalClusterAgentUtil, nil
+}
+
+func (c *ClusterAgentUtil) init() error {
+	var err error
+
+	c.clusterAgentAPIEndpoint, err = getClusterAgentEndpoint()
+	if err != nil {
+		return err
+	}
+
+	authToken := config.Datadog.GetString("cluster_agent_auth_token")
+	if authToken == "" {
+		return fmt.Errorf("empty cluster_agent_auth_token")
+	}
+	if len(authToken) < clusterAgentAuthTokenMinLen {
+		return fmt.Errorf("need at least a length of %d for cluster_agent_auth_token: %d", clusterAgentAuthTokenMinLen, len(authToken))
+	}
+
+	c.clusterAgentAPIRequestHeaders = &http.Header{}
+	c.clusterAgentAPIRequestHeaders.Set(authorizationHeaderKey, fmt.Sprintf("Bearer %s", authToken))
+
+	// TODO remove insecure
+	c.clusterAgentAPIClient = GetClient(false)
+	c.clusterAgentAPIClient.Timeout = 2 * time.Second
+
+	return nil
+}
+
+// getClusterAgentEndpoint provides a validated https endpoint from:
+// 1. configuration key "cluster_agent_url" like https
+// add the https prefix if the scheme isn't specified
+// 2. environment variables associated with "cluster_agent_kubernetes_service_name"
+// ${dcaServiceName}_SERVICE_HOST and ${dcaServiceName}_SERVICE_PORT
+func getClusterAgentEndpoint() (string, error) {
+	const configDcaURL = "cluster_agent_url"
+	const configDcaSvcName = "cluster_agent_kubernetes_service_name"
+
+	dcaURL := config.Datadog.GetString(configDcaURL)
+	if dcaURL != "" {
+		if strings.HasPrefix(dcaURL, "http://") {
+			return "", fmt.Errorf("cannot get cluster agent endpoint, not a https scheme: %s", dcaURL)
+		}
+		if strings.Contains(dcaURL, "://") == false {
+			log.Tracef("adding https scheme to %s: https://%s", dcaURL, dcaURL)
```

Comment:
I think `adding https scheme to %s` suffices

Reply:
If the user specify `tcp://` we insert involuntary `https://tcp://...`
I'll see how to improve this part

Reply:
See [tcp](https://github.com/DataDog/datadog-agent/pull/1127/files#diff-caa70174faffdcf870b85dd14e28883dR153) error handling